File: MediaData/src/main/java/com/tool/mediadata/utils/FolderLoader.java
package com.tool.mediadata.utils;

import android.content.Context;
import android.provider.MediaStore;
import android.text.TextUtils;

import com.tool.mediadata.entity.Folder;
import com.tool.mediadata.entity.Music;

import java.io.File;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * desc:
 * *
 * user: xujj
 * time: 2022/8/26 14:55
 **/
public class FolderLoader {

    /**
     * 通过歌曲列表获取文件夹列表
     *
     * @param musics //按路径排序的歌曲列表
     */
    private static List<Folder> getFolderList(List<Music> musics, String name) {
        List<Folder> folderList = new ArrayList<>();
        Map<String, Folder> folderMap = new LinkedHashMap<>();
        try {
            for (Music music : musics) {
                File file = new File(music.getData()).getParentFile();
                if (file != null) {
                    String folderName = file.getName();
                    String path = file.getPath();

                    Folder folder = folderMap.get(path);
                    if (folder == null) {
                        folder = new Folder(folderName, path, 1);
                        folderMap.put(path, folder);
                    } else {
                        folder.setMusicCount(folder.getMusicCount() + 1);
                        folderMap.put(path, folder);
                    }
                }
            }
            Set<String> set = folderMap.keySet();
            for (String path : set) {
                Folder folder = folderMap.get(path);
                if (folder != null) {
                    if (TextUtils.isEmpty(name)) {
                        folderList.add(folder);
                    } else {
                        if (folder.getName().toLowerCase().contains(name.toLowerCase())) {
                            folderList.add(folder);
                        }
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return folderList;
    }

    public static List<Folder> getFolderList(Context context, String name) {
        List<Music> musics = MusicLoader.getMusicList(context, null, MediaStore.Audio.Media.DATA, null);
        return getFolderList(musics, name);
    }

    public static List<Folder> getFolderList(Context context) {
        return getFolderList(context, null);
    }
}

File: MediaData/src/main/java/com/tool/mediadata/utils/FolderLoader.kt
package com.tool.mediadata.utils

import android.content.Context
import com.tool.mediadata.MediaConfig
import com.tool.mediadata.entity.Folder
import com.tool.mediadata.entity.Music
import java.io.File

/**
 * desc:
 **
 * user: xujj
 * time: 2023/10/26 14:53
 **/
object FolderLoader {

    fun getFolderListBase(
        musics: List<Music>?,
        name: String? = null,
        sortOrder: String? = MediaConfig.getInstance().folderSortOrder
    ): MutableList<Folder> {
        if (musics.isNullOrEmpty()) return ArrayList()

        val folderList = ArrayList<Folder>()
        val folderMap = LinkedHashMap<String, Folder>()
        try {
            for (music in musics) {
                val file: File? = File(music.data).parentFile
                if (file != null) {
                    val folderName = file.name
                    val path = file.path
                    var folder: Folder? = folderMap[path]
                    if (folder == null) {
                        folder = Folder(folderName, path, 1)
                        folderMap[path] = folder
                    } else {
                        folder.musicCount = folder.musicCount + 1
                        folderMap[path] = folder
                    }
                }
            }
            val set: Set<String> = folderMap.keys
            for (path in set) {
                val folder: Folder? = folderMap[path]
                if (folder != null) {
                    if (name.isNullOrEmpty()) {
                        folderList.add(folder)
                    } else {
                        if (folder.name.lowercase().contains(name.lowercase())) {
                            folderList.add(folder)
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        when (sortOrder) {
            SortOrder.FolderSortOrder.FOLDER_A_Z -> {
                folderList.sortBy {
                    it.name
                }
            }

            SortOrder.FolderSortOrder.FOLDER_Z_A -> {
                folderList.sortByDescending {
                    it.name
                }
            }

            SortOrder.FolderSortOrder.FOLDER_MUSIC_COUNT -> {
                folderList.sortBy {
                    it.musicCount
                }
            }

            SortOrder.FolderSortOrder.FOLDER_MUSIC_COUNT_DESC -> {
                folderList.sortByDescending {
                    it.musicCount
                }
            }
        }
        return folderList
    }

    fun getFolderList(
        context: Context,
        name: String? = null,
        sortOrder: String? = MediaConfig.getInstance().folderSortOrder
    ): MutableList<Folder> {
        val musicList = MusicLoader.getAllMusicList(context)
        return getFolderListBase(musicList, name, sortOrder)
    }
}